### DISCUSSION TITLE
Reconsider allowing global offsets for calendar entities

### DISCUSSION THREAD

**Discussion by decompil3d** [home-assistant/architecture#700] [Entity Models] [answered]
Hi folks. About 2 years ago, home-assistant/core#29934 was opened to allow setting global event offsets for calendars. At the time, @balloob closed the PR saying,

> We should not merge this. We should remove all offsets from the calendar integration instead. Integrations are either responsible for delivering data from an API as-is, or process data in Home Assistant and show that.

I'd like to request that we reconsider this stance and re-open the PR. I'd argue that the solution in the PR is, in fact, a means of allowing HA to process as-is data from an API.

Since calendar entities only allow loading the single current or upcoming event, there is no other mechanism to reason over upcoming events (at least as far as I know). An alternative approach could be to provide some sort of template helper to be able to retrieve a later event from a calendar entity. The HA REST API already seems to expose arbitrary events from calendars via the `/calendars/<calendar entity>` route, so perhaps this is a reasonable path forward.

Either way, [there](https://github.com/home-assistant/core/pull/29934) [is](https://github.com/home-assistant/core/pull/50856) [clear](https://github.com/home-assistant/core/pull/47049) [interest](https://github.com/home-assistant/core/pull/44831) [in](https://github.com/home-assistant/core/pull/38201) being able to reason over future calendar events in HA. I'd appreciate any consideration on mechanisms to enable this scenario. Thank you!

**Comment by MartinHjelmare** [comment, marked as answer]
I think we should build a trigger platform for the calendar integration so that the user can register triggers that will fire on matches against calendar event data and time relative to the event.

This will remove the need for the currently available offset system.

**Reply by decompil3d** [reply]
To make sure that I understand, are you suggesting using [this process](https://developers.home-assistant.io/docs/device_automation_trigger)? If so, would it make sense to attach this at the root `calendar` component or on the `google` calendar component?

**Reply by MartinHjelmare** [reply]
No, not device triggers. An integration can create a trigger platform to support custom triggers. See eg the zwave_js integration.

I think we should try to make it standardized on the calendar integration so that all integrations with calendar platforms can leverage it with minimal work. We already have a standard interface to fetch calendar events for a time period. That may be enough to let us implement it.

**Reply by decompil3d** [reply]
Thanks for clarifying. Is this something you'd like me to attempt to contribute or something better left to a maintainer or core contributor? I'm happy to help, though I'm only a few PRs deep into HA core development so far.

**Reply by MartinHjelmare** [reply]
Let's start with describing the trigger config schema and then what trigger data we want available in the automation when it has fired.

This will help us see what will be needed to build the idea.

**Reply by decompil3d** [reply]
Sounds good. For my needs, I'd like to be able to trigger an automation _n_ minutes ahead of the next event. I would likely also want the ability to limit my automation to events that are "busy" on the calender (i.e. not "free") and also to ensure that there's not a current event in progress. I could do that using conditions though.

**Reply by decompil3d** [reply]
Here's my initial stab at a config schema and trigger data schema:

### Trigger config schema

```yaml
# Required. Entity ID of calendar to search.
calendar: 'calendar.personal'

# Required. String to search for in the title and description of events.
search: 'search string'

# Optional. Offset from the event start time to trigger at.
# Same data type that's currently used for the offset mechanism in our calendar integrations
offset: '-00:10'

# Optional. Whether to exclude all-day events. Default false.
exclude_all_day: false
```

#### Open questions

- Should `calendar` support multiple calendars? Maybe not an MVP feature.
- Should the `search` field look at location fields too? Should we allow specifying which fields to search separately? Or perhaps allow separate fields for searching each part? I'd suggest that we just have this single `search` field for the MVP.

### Trigger data schema

- All attributes that would be available in the calendar sensor when the event is active (e.g. `message`, `description`, `all_day`, `start_time`, etc.)

### Implementation thoughts

In terms of implementation, I'm thinking this could be done by running a periodic update that reads events up to some amount of time in the future (e.g. 12 hours ahead, or whatever the update frequency is). For each one, check if it matches the `search`. If so, retrieve its `start_time`, apply the `offset`, and then set a time trigger for that resulting time. Store the time trigger object in a `List` for later reference. When updating, clear all existing time triggers and recreate as needed.

**Open question:** I'm not sure how expensive it is to create time triggers. If we need to avoid deleting/recreating them each update, we could do something more complex like storing a `Dict` of event ID -> time trigger object and re-use existing triggers when they don't need to change.

@MartinHjelmare Am I on the right track here? Any thoughts?

**Reply by allenporter** [reply]
I'm definitely not an expert in triggers at all, but am chiming in since i've been staring at the code in google calendar lately, doing some cleanup.

For the config schema: It seems like there is a need for an `event` type of something like `START` and `END` that are triggered at the start or end of the event?

For your questions, for what it is worth:
- I agree that a single calendar in a  trigger seems fine to me. You can add more triggers if you want to match additional calendars.
-  I agree w/ keeping search simple.

I have a hunch that exclude_all_day could somehow be made more elegant, but don't have a suggestion.

**Reply by allenporter** [reply]
My impression is that part of the motivation for this proposal is to also remove the existing entities that act as binary sensors? 

Today it looks like `google` creates one entity per search.  Would it instead have one entity per calendar? The UI would then change to show all events on the calendar, rather than just those that match the search criteria, since the search critiera moves to a trigger.

Looking at`todoist`, this schema seems like it could be a step towards replacing the custom_projects semantics. However, I'm not sure what would happen  to `label` and `include_projects`. Is `label` just `search`? is `include_projects` just a separate calendar id to match?

Regarding the implementation thoughts:
This seems like it can be modeled similarly to the sun trigger, which works similarly to what you said, but has an object per trigger responsible for rescheduling itself (`SunListener`) based on the next instance of the sun event (with an offset) -- however, I believe the next sun event is pretty static, where as here the calendar may change and the trigger may need to reschedule itself, as you said.  But i like that Sun has one object per trigger, and it is responsible for rescheduling itself, rather than adding/removing all the trigger timers each refresh.

**Reply by MartinHjelmare** [reply]
I think all comments so far sound as we're going in the right direction with the discussion. Please keep thinking and discussing. This is not a simple problem so we'll need a careful plan before proceeding.

Regarding the GUI, we have the calendar view that shows all events from all calendars so entities are not needed for that purpose.

If the automation needs can be fulfilled by the new calendar trigger, the entities serving as automation targets may become obsolete. We may still want an entity per calendar to record the calendars, allow using the entity registry etc. We should think about what state the calendar entity, if it's kept, should express. Just next event time?

**Reply by allenporter** [reply]
@decompil3d i'm curious which calendar integration you're using? caldav or google?

Implementation wise, i think this was interesting discussion because `google` doesn't currently have an entity per calendar at the moment, but it seems like things could get much simpler if it did. Today it seems like it runs a separate query for every `search` which is a separate entity.  If there was a simple entity responsible for fetching all events.

Here is an alternative trigger / event approach that I don't think works.  It would be nice and simple to publish a new event `calendar_event` with:
- `calendar_event_type`: `CALENDAR_EVENT_STARTED` or `CALENDAR_EVENT_ENDED` (or something shorter)
- `calendar_id`: The id of the calendar. (Is this really `calendar_entity_id`?)
- `calendar_event_data`: A dict with the payload of the event (start, stop, title, summary, all_day, whatever custom fields the integration supports)
and then  implement the calendar trigger on top of those events with very basic user friendly options, but still allow more complex triggers based on the event payload data if users want to extend something more complicated.  My impression is this is similar to how device triggers are implemented.  

However, I realize this approach does not work with offsets.  I am just mentioning this in case it sparks another idea.

**Reply by decompil3d** [reply]
I use the `google` integration although I've dabbled with the iCalendar one in HACS as well. Offset is important to me as I want to notify ahead of events typically.

**Reply by allenporter** [reply]
I've been thinking about this proposal more and would like to suggest a modification, and remove a couple of the fields from the trigger config schema.

## Trigger config schema

The trigger config schema only contains inputs necessary for starting or ending an event:
```
# Required. Entity ID of calendar to search.
calendar: 'calendar.personal'

# Indicates if the trigger fires when the event starts or ends
event: 'started' or 'ended'

# Optional. Offset from the event start time to trigger at.
# Same data type that's currently used for the offset mechanism in our calendar integrations
offset: '-00:10'
```

## Trigger data schema

The payload for the trigger may contain integration specific data that is used for evaluating conditions in automations, and support much more extensive information.  For example, the integration may include an `event` dictionary with additional fields such as `start_time`, `end_time`, `all_day`, `summary`, `description`, etc that an automation may used as inputs to a condition.

## Rationale 

Google calendar currently supports sensor attributes like `search` or `all_day` as part of the entity/event filtering, and it can continue to do that  -- or not and move them to be implemented by users as part of their condition.  With respect to trigger schema, it doesn't really matter. It may be helpful for integrations to use common conventions for the data payloads (e.g. they copy what google calendar does), but like any other type of trigger payload for an integration, they often have specifics not supported by other integrations. Basically, this is all detail that now can be considered out of scope for the trigger itself.

**Reply by MartinHjelmare** [reply]
Can the offset item be used for the event end too, if `event` is set to `ended`?

**Reply by allenporter** [reply]
> Can the offset item be used for the event end too, if `event` is set to `ended`?

I deleted my prior answer since I misread your question and answered a totally different question. (Read too fast on mobile)

Yes, with this schema, offset would work for end events too.

**Reply by allenporter** [reply]
@decompil3d Do you have thoughts on a slimmed down trigger config (with offsets) keeping the existing event filtering logic in the integration like it is today?  I think this can be a long term solution, but also seems like a good first step that we can revisit later if we change our minds and it seems like we have more flexibility keeping the extra fields in the trigger data, since its integration specific (vs the trigger schema seems like a bigger commitment).

**Reply by decompil3d** [reply]
@allenporter I am good with that. I believe that @MartinHjelmare  had asked for the filter/search logic in the trigger which is why I included it. But my main interests are the start time and offset.

**Reply by allenporter** [reply]
Thinking through the implementation, I think we may want one data stream of events per calendar. An individual attached trigger  would grab the stream of events, apply its offset, then keep handy the list of events to iterate through (via point in time alarms) and fire actions for. We could only support offsets + or - 12-24 hours from the event start (or end) to limit the data fetched.

Providing the data stream, I see two options:
- The trigger platform has a schedule fetcher for each calendar, and it is refreshed outside of the normal entity update schedule
- The entity itself keeps the state of al the events, and the integration is in charge of driving the refresh schedule

While it seems less typical, I am leaning towards option #1 as having the entity hold on to a list of events seems non-typical, however I am very open to suggestion. I could see making the section option work by expanding the API of the calendar, for example, but it wasn't clear to me it was worth it yet.

As for entity state, i think next available event time makes sense, but really, i'm not seeing any use for any state on the entity if we want to leave it alone even. Changing it, tough, could drive people to use the triggers.

@MartinHjelmare are there other topics we should think through before implementation?

**Reply by MartinHjelmare** [reply]
I think we can use the existing interface to fetch calendar events, and then we can put the schedule fetcher in the trigger platform.

Maybe just schedule one point in time alarm for the next event per trigger? On every fetch match the next event data against the existing scheduled event data, and if it's not the same reschedule it.

The three main points to think about are:

1. Trigger config schema
2. Trigger scheduling and event refresh
3. Entity model

For 1. I think the slimmed down version with only three items (entity_id, event, offset) sounds good.

We're talking about 2. now.

For 3. I think we can leave it as it is for now in this discussion.

**Reply by allenporter** [reply]
On #2 what you said matches what I was thinking also after some prototyping. (trigger does the refresh itself, schedules a callback for the next event, no entity API changes needed.)

There are tradeoffs to make on the spectrum of code simplicity vs handling user expectations well about corner cases, and then implications on efficiency.

When thinking about the data fetching to schedule the next trigger time there are a few cases to consider:
- a) after a trigger fires, we determine the next event. Note also there may be multiple events with the same start time and we'd want to accurately fire for each of them.
- b) a new event is added to the calendar, and it's the next event, so we need to fix the next refresh time. Fairly rare, but still we want to consider the minimum time for detection (today 15 minutes so maybe we stick with that)
- c) End event trigger handling is slightly more complex than start time handling, given the API scan date ranges are in terms of start/end of the start of the event. The trigger data fetch would look a bit in the past to ensure it sees enough events that are sending soon.


When thinking about cloud based calendars and making this performant, ideally we don't have multiple things polling the API. However maybe we say that in order to keep the core simple, we push performance issues to be the responsibility of the integration? e.g. it implements the existing APIs with a local backed event cache or sync down from cloud if we care about avoiding duplicate polling rpc calls. Otherwise we have to push coordination into calendar, or just punt and poll twice.

Curious to hear more feedback.

**Reply by MartinHjelmare** [reply]
Yes, I think that each integration should be responsible for optimizing their api calls as needed. They know what they need to do.

**Reply by allenporter** [reply]
@frenck I would appreciate your input on this discussion thread given your recent experiences with calendar, when you have a moment, though this is not urgent.

**Reply by balloob** [reply]
I 👍 this proposal

**Reply by MartinHjelmare** [reply]
Let's go with this for the MVP.

**Comment by allenporter** [comment]
Another opposite approach could be to embrace the approach of how google calendar works with separate entities per search term.  Searches are baked into the entity.  Triggers would then only be in terms of start and stop + offsets, without extra filtering functionality. That can be left up to the integration.

The down side of doing this is if there are common filtering things we want to support across all events, then they all need to implement it separately.  Likely we don't want to go with this approach.

Typically something like `search` would be an integration specific feature hidden behind the API supporting it. However, i think having search support common across all integrations seems like it could be simpler as the integration only needs to manage a single schedule rather than multiple schedules for each term making many queries.

**Comment by markus99** [comment]
I wrote about a related issue here https://github.com/home-assistant/core/issues/66666

In my opinion, while the triggers on multiple events would be helpful (a infinitely better than we have now [they're ignored]), they'd potentially not solve the entire problem.  

For example, birthdays are 'all day' events, which I believe in this solution means they'd trigger at midnight - which might be too early for someone.  Personally, I have a Slack notification that triggers at 7:30am posting a happy birthday announcement for our company.  Triggering at midnight would be too early and I'd have to use these proposed triggers to update numerous input_text. helpers to then be referenced conditionally using my time-based-trigger at 7:30am.

My request would be to return, as an array, the multiple events in the calendar entity itself.   The calendar entity could, by default, return an array of ALL events during the current day - for the user to reference as they like.

As a side thought, could also, optionally, create additional calendar entities for future days in calendar setup.  This would help with the use-case presented here: https://community.home-assistant.io/t/wth-proper-calendar-agenda-that-tells-me-my-appointments-for-today-or-tomorrow/219994/5

**Reply by MartinHjelmare** [reply]
Why wouldn't you be able to use the offset to trigger the automation at the wanted time for all day events?
